app/src/main/java/app/revanced/integrations/shared/checks/Check.java

```diff
@@ -0,0 +1,123 @@
+                    .setTitle(str("revanced_check_environment_failed_title"))
+                    .setPositiveButton(
+                            str("revanced_check_environment_dialog_open_official_source_button"),
+                            (dialog, which) -> {
+                                final var intent = new Intent(Intent.ACTION_VIEW, GOOD_SOURCE);
+                                intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
+                                context.startActivity(intent);
+                            }
+                    ).setNegativeButton(
+                            str("revanced_check_environment_dialog_ignore_button"),
+                            (dialog, which) -> {
+                                final int current = Settings.CHECK_ENVIRONMENT_WARNING_ISSUED_COUNT.get();
+                                Settings.CHECK_ENVIRONMENT_WARNING_ISSUED_COUNT.save(current + 1);
+
+                                dialog.dismiss();
+                            }
+                    );
+
+    private final String failedReasonStringKey;
+
+    /**
+     * @param failedReasonStringKey The string key of the reason why the check failed to be shown to the user.
+     */
+    Check(String failedReasonStringKey) {
+        this.failedReasonStringKey = failedReasonStringKey;
+    }
+
+    protected abstract boolean run();
+
+    static boolean shouldRun() {
+        return Settings.CHECK_ENVIRONMENT_WARNING_ISSUED_COUNT.get() < 2;
+    }
+
+    static void disableForever() {
+        Logger.printDebug(() -> "Environment checks disabled forever.");
+
+        Settings.CHECK_ENVIRONMENT_WARNING_ISSUED_COUNT.save(Integer.MAX_VALUE);
+    }
+
+    @SuppressLint("NewApi")
+    static void issueWarning(Context context, Check... failedChecks) {
+        final var reasons = new StringBuilder();
+
+        reasons.append("<ul>");
+
+        for (var check : failedChecks) {
+            reasons.append("<li>").append(str(check.failedReasonStringKey));
+        }
+
+        reasons.append("</ul>");
+
+        final var finalMessage = Html.fromHtml(
+                String.format(str("revanced_check_environment_failed_message"), reasons),
+                FROM_HTML_MODE_COMPACT
+        );
+        AlertDialog dialog = CHECK_FAILED_DIALOG_BUILDER.apply(context)
+                .setMessage(finalMessage).create();
+
+        var dismissButton = dialog.getButton(DialogInterface.BUTTON_NEGATIVE);
+        // TODO: setEnabled called on null? If this line is commented out getting:
+        //  Unable to add window -- token null is not valid; is your activity running?
+        dismissButton.setEnabled(false);
+
+        dialog.show();
+
+        // Use a delay to allow the activity to finish initializing.
+        // Otherwise, if device is in dark mode the dialog is shown with wrong color scheme.
+        Utils.runOnMainThreadDelayed(getCountdownRunnable(dismissButton), 100);
```

Comment:
Seems like this doesn't work but works for GmsCoreSupport @LisoUseInAIKyrios should we increase the delay?